### PR TITLE
Ravel model for the illusion of 2D tables

### LINKED ISSUE #168: WCSes using a 2D Lookup table for pointing do not support inverse transform
When constructing a `VaryingCelestialTransform2D` model we [currently use `Tabular2D`](https://bitbucket.org/dkistdc/dkist-inventory/src/359e6b0c9f28586d821874b4ecc60d4449e2633d/dkist_inventory/transforms.py#lines-192:194) which does not support an inverse transform even when it's all singularly valued.

The approach to fixing this is probably to make a subclass of `Tabular2D` which does return an inverse when all the points in the table are singularly valued. This probably should be upstreamed to Astropy or else we will have to maintain it forever.

See also DCS-2680

### PR DESCRIPTION
**Pull request by SolarDrew**
Closes #168 (once dkist-inventory is updated to use these changes).

Adds a new model to take a 2D index and return the corresponding correct index for a 1D array, and the inverse model for the reverse operation. To be used as a compound with `Tabular1D` so that it looks like a `Tabular2D` but the compound model can still be inverted.

Should be straightforward to expand to 3D also.

### CODE REVIEW

File: dkist/io/asdf/entry_points.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def get_extensions():
                                    converters=dkist_converters),
         ManifestExtension.from_uri("asdf://dkist.nso.edu/manifests/dkist-1.0.0",
                                    converters=dkist_converters),
-        ManifestExtension.from_uri("asdf://dkist.nso.edu/manifests/dkist-wcs-1.0.0",
+        ManifestExtension.from_uri("asdf://dkist.nso.edu/manifests/dkist-wcs-1.1.0",
```
Comment (Cadair) [inline dkist/io/asdf/entry_points.py:51]: We need both here, not just the new one.

File: dkist/io/asdf/resources/manifests/dkist-wcs-1.1.0.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+%YAML 1.1
+---
+id: asdf://dkist.nso.edu/dkist/manifests/dkist-wcs-1.0.0
+extension_uri: asdf://dkist.nso.edu/dkist/extensions/dkist-wcs-1.0.0
```
Comment (Cadair) [inline dkist/io/asdf/resources/manifests/dkist-wcs-1.1.0.yaml:4]: 1.1

**Comment by Cadair** [comment]
Fixed that for you :wink:

**Comment by Cadair** [comment]
I am in your code, finding your bugs.

I made an asdf, and the first thing I noticed is this:
```
NotImplementedError: The "separable" property is not defined for model Ravel
```
<details>

```
In [14]: print(ds)
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[14], line 1
----> 1 print(ds)

File ~/Git/DKIST/dkist/dkist/dataset/dataset.py:224, in Dataset.__str__(self)
    223 def __str__(self):
--> 224     return dataset_info_str(self)

File ~/Git/DKIST/dkist/dkist/dataset/utils.py:92, in dataset_info_str(ds)
     83 s += (' ' * world_dim_width + '  ' +
     84         ('{0:^' + str(wcs.pixel_n_dim * 5 - 2) + 's}').format('Pixel Dim') +
     85         '\n')
     87 s += (('{0:' + str(world_dim_width) + 's}').format('World Dim') +
     88         ''.join(['  ' + ('{0:' + str(pixel_dim_width) + 'd}').format(ipix)
     89                 for ipix in range(wcs.pixel_n_dim)]) +
     90         '\n')
---> 92 matrix = wcs.axis_correlation_matrix[::-1, ::-1]
     93 matrix_str = np.empty(matrix.shape, dtype='U3')
     94 matrix_str[matrix] = 'yes'

File ~/.virtualenvs/dkist/lib/python3.11/site-packages/gwcs/api.py:249, in GWCSAPIMixin.axis_correlation_matrix(self)
    239 @property
    240 def axis_correlation_matrix(self):
    241     """
    242     Returns an (`~BaseLowLevelWCS.world_n_dim`,
    243     `~BaseLowLevelWCS.pixel_n_dim`) matrix that indicates using booleans
   (...)
    247     would be `True` and all other entries `False`.
    248     """
--> 249     return separable.separability_matrix(self.forward_transform)

File ~/Git/astropy/astropy/modeling/separable.py:98, in separability_matrix(transform)
     96 if transform.n_inputs == 1 and transform.n_outputs > 1:
     97     return np.ones((transform.n_outputs, transform.n_inputs), dtype=np.bool_)
---> 98 separable_matrix = _separable(transform)
     99 separable_matrix = np.where(separable_matrix != 0, True, False)
    100 return separable_matrix

File ~/Git/astropy/astropy/modeling/separable.py:307, in _separable(transform)
    305 elif isinstance(transform, CompoundModel):
    306     sepleft = _separable(transform.left)
--> 307     sepright = _separable(transform.right)
    308     return _operators[transform.op](sepleft, sepright)
    309 elif isinstance(transform, Model):

File ~/Git/astropy/astropy/modeling/separable.py:302, in _separable(transform)
    287 def _separable(transform):
    288     """
    289     Calculate the separability of outputs.
    290 
   (...)
    299         Each element represents the separablity of the corresponding output.
    300     """
    301     if (
--> 302         transform_matrix := transform._calculate_separability_matrix()
    303     ) is not NotImplemented:
    304         return transform_matrix
    305     elif isinstance(transform, CompoundModel):

File ~/Git/DKIST/dkist/dkist/wcs/models.py:672, in CoupledCompoundModel._calculate_separability_matrix(self)
    670 def _calculate_separability_matrix(self):
    671     sepleft = separable._separable(self.left)
--> 672     sepright = separable._separable(self.right)
    674     noutp = separable._compute_n_outputs(sepleft, sepright)
    676     cleft = np.zeros((noutp, sepleft.shape[1]))

File ~/Git/astropy/astropy/modeling/separable.py:306, in _separable(transform)
    304     return transform_matrix
    305 elif isinstance(transform, CompoundModel):
--> 306     sepleft = _separable(transform.left)
    307     sepright = _separable(transform.right)
    308     return _operators[transform.op](sepleft, sepright)

File ~/Git/astropy/astropy/modeling/separable.py:306, in _separable(transform)
    304     return transform_matrix
    305 elif isinstance(transform, CompoundModel):
--> 306     sepleft = _separable(transform.left)
    307     sepright = _separable(transform.right)
    308     return _operators[transform.op](sepleft, sepright)

File ~/Git/astropy/astropy/modeling/separable.py:310, in _separable(transform)
    308     return _operators[transform.op](sepleft, sepright)
    309 elif isinstance(transform, Model):
--> 310     return _coord_matrix(transform, "left", transform.n_outputs)

File ~/Git/astropy/astropy/modeling/separable.py:200, in _coord_matrix(model, pos, noutp)
    198         mat[-model.n_outputs :, -model.n_inputs :] = m
    199     return mat
--> 200 if not model.separable:
    201     # this does not work for more than 2 coordinates
    202     mat = np.zeros((noutp, model.n_inputs))
    203     if pos == "left":

File ~/Git/astropy/astropy/modeling/core.py:1589, in Model.separable(self)
   1587 if self._separable is not None:
   1588     return self._separable
-> 1589 raise NotImplementedError(
   1590     'The "separable" property is not defined for '
   1591     f"model {self.__class__.__name__}"
   1592 )

NotImplementedError: The "separable" property is not defined for model Ravel
```

</details>

which breaks the repr / str of the `dataset` class amongst other things. I think it should be a simple case of setting [`_separable = False`](https://github.com/DKISTDC/dkist/blob/main/dkist/wcs/models.py#LL88).

**Comment by Cadair** [comment]
I added the fix, because I needed it for further testing.

**Comment by SolarDrew** [comment]
Well I thought I'd found and fixed that bug myself but thanks I guess :laughing:

File: changelog/227.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Add a new model to take a 2D index and return the corresponding correct index for a 1D array, and the inverse model for the reverse operation.
+To be used as a compound with Tabular1D so that it looks like a Tabular2D but the compound model can still be inverted.
+This will be relevant and useful in the dkist-inventory package but users are unlikely to need to use it.
```
Comment (Cadair) [inline changelog/227.bugfix.rst:3]:
```suggestion
```

File: dkist/wcs/models.py [hunk 4]
```diff
@@ -733,3 +734,54 @@ def varying_celestial_transform_from_tables(crpix: Union[Iterable[float], u.Quan
...
+    array_shape = (0, 0)
+    n_inputs = 2
+    n_outputs = 1
+    _separable = False
+
+    def __init__(self, array_shape, order='C', **kwargs):
+        super().__init__(**kwargs)
+
+        self.array_shape = tuple(array_shape)
+        self.order = order  # TODO: Validate order is either C or F
```
Comment (Cadair) [inline dkist/wcs/models.py:749]:
```suggestion
        if order not in ("C", "F"):
            raise ValueError("order kwarg must be one of 'C' or 'F'")
        self.order = order
```

File: dkist/wcs/models.py [hunk 5]
```diff
@@ -733,3 +734,54 @@ def varying_celestial_transform_from_tables(crpix: Union[Iterable[float], u.Quan
...
+    array_shape = (0, 0)
+    n_inputs = 1
+    n_outputs = 2
+    _separable = False
+
+    def __init__(self, array_shape, order='C', **kwargs):
+        super().__init__(**kwargs)
+
+        self.array_shape = array_shape
+        self.order = order  # TODO: Validate order is either C or F
```
Comment (Cadair) [inline dkist/wcs/models.py:776]:
```suggestion
        if order not in ("C", "F"):
            raise ValueError("order kwarg must be one of 'C' or 'F'")
        self.order = order
```

**Review by Cadair — APPROVED** [review]

**Pull request merged**